File: api/upsolve-data.js
import { MongoClient } from 'mongodb';
import { getAuthFromRequest } from './auth-shared.js';

const uri = process.env.MONGODB_URI;
const dbName = process.env.DB_NAME || 'skilltree';

let cachedClient = null;

async function getDb() {
    if (cachedClient && cachedClient.topology && cachedClient.topology.isConnected()) {
        return cachedClient.db(dbName);
    }
    const client = new MongoClient(uri);
    await client.connect();
    cachedClient = client;
    return client.db(dbName);
}

function sendJson(res, status, body) {
    res.statusCode = status;
    res.setHeader('Content-Type', 'application/json; charset=utf-8');
    res.setHeader('Cache-Control', 'no-store');
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');
    res.end(JSON.stringify(body));
}

async function readJsonBody(req) {
    if (req.body && typeof req.body === 'object') return req.body;
    if (typeof req.body === 'string') {
        try { return JSON.parse(req.body); } catch { return null; }
    }
    const raw = await new Promise((resolve, reject) => {
        let data = '';
        req.on('data', (chunk) => (data += chunk));
        req.on('end', () => resolve(data));
        req.on('error', reject);
    });
    if (!raw) return null;
    try { return JSON.parse(raw); } catch { return null; }
}

const EMPTY_DATA = { todo: [], solved: [] };

export default async function handler(req, res) {
    const method = req.method || 'GET';

    if (method === 'OPTIONS') {
        return sendJson(res, 200, { ok: true });
    }

    if (!uri) {
        return sendJson(res, 500, { error: 'MONGODB_URI not configured' });
    }

    const auth = getAuthFromRequest(req);
    if (!auth || !auth.userId) {
        return sendJson(res, 401, { error: 'Authentication required' });
    }

    try {
        const db = await getDb();
        const col = db.collection('upsolve_data');

        if (method === 'GET') {
            const urlObj = new URL(req.url, `https://${req.headers.host || 'localhost'}`);
            const handle = urlObj.searchParams.get('handle');
            if (!handle) {
                return sendJson(res, 200, EMPTY_DATA);
            }
            const doc = await col.findOne({ _id: handle });
            if (doc) {
                const { _id, ...data } = doc;
                return sendJson(res, 200, data);
            }
            return sendJson(res, 200, EMPTY_DATA);
        }

        if (method === 'POST') {
            const payload = await readJsonBody(req);
            if (!payload || typeof payload !== 'object') {
                return sendJson(res, 400, { error: 'Invalid JSON' });
            }

            const handle = payload.handle;
            if (!handle || typeof handle !== 'string') {
                return sendJson(res, 400, { error: 'handle is required' });
            }

            const action = payload.action;

            if (action === 'add_todo') {
                const problem = payload.problem;
                if (!problem) return sendJson(res, 400, { error: 'problem is required' });

                await col.updateOne(
                    { _id: handle },
                    {
                        $push: { todo: problem },
                        $set: { savedAt: new Date().toISOString() }
                    },
                    { upsert: true }
                );
                return sendJson(res, 200, { ok: true });
            }

            if (action === 'mark_solved') {
                const contestId = payload.contestId;
                const index = payload.index;
                if (!contestId || !index) return sendJson(res, 400, { error: 'contestId and index required' });

                const doc = await col.findOne({ _id: handle });
                if (!doc) return sendJson(res, 404, { error: 'No upsolve data found' });

                const todoItem = (doc.todo || []).find(p => p.contestId === contestId && p.index === index);
                if (!todoItem) return sendJson(res, 404, { error: 'Problem not found in todo' });

                await col.updateOne(
                    { _id: handle },
                    {
                        $pull: { todo: { contestId, index } },
                        $push: { solved: { ...todoItem, solvedAt: new Date().toISOString() } },
                        $set: { savedAt: new Date().toISOString() }
                    }
                );
                return sendJson(res, 200, { ok: true });
            }

            if (action === 'remove_todo') {
                const contestId = payload.contestId;
                const index = payload.index;
                if (!contestId || !index) return sendJson(res, 400, { error: 'contestId and index required' });

                await col.updateOne(
                    { _id: handle },
                    {
                        $pull: { todo: { contestId, index } },
                        $set: { savedAt: new Date().toISOString() }
                    }
                );
                return sendJson(res, 200, { ok: true });
            }

            return sendJson(res, 400, { error: 'Unknown action. Use add_todo, mark_solved, or remove_todo' });
        }

        return sendJson(res, 405, { error: 'Method Not Allowed' });
    } catch (err) {
        console.error('upsolve-data error:', err);
        return sendJson(res, 500, { error: 'Database error' });
    }
}

File: api/upsolve_data.py
"""
Upsolve data API — MongoDB CRUD for upsolve todo/solved lists.
Port of api/upsolve-data.js
"""

import os
from datetime import datetime, timezone
from flask import Flask, request
from flask_cors import CORS
from pymongo import MongoClient
from api.auth_utils import get_auth_from_request, send_cors_json

app = Flask(__name__)
CORS(app, resources={r"/api/*": {"origins": "*"}})

MONGODB_URI = os.environ.get('MONGODB_URI', '')
DB_NAME = os.environ.get('DB_NAME', 'skilltree')

_client = None

EMPTY_DATA = {'todo': [], 'solved': []}


def get_db():
    global _client
    if not MONGODB_URI:
        return None
    if _client is not None:
        try:
            _client.admin.command('ping')
            return _client[DB_NAME]
        except Exception:
            _client = None
    _client = MongoClient(MONGODB_URI, serverSelectionTimeoutMS=10000)
    _client.admin.command('ping')
    return _client[DB_NAME]


@app.route('/api/upsolve-data', methods=['GET', 'POST', 'OPTIONS'])
def upsolve_data():
    if request.method == 'OPTIONS':
        return send_cors_json({'ok': True})

    if not MONGODB_URI:
        return send_cors_json({'error': 'MONGODB_URI not configured'}, 500)

    auth = get_auth_from_request(request)
    if not auth or not auth.get('userId'):
        return send_cors_json({'error': 'Authentication required'}, 401)

    try:
        db = get_db()
        col = db.upsolve_data

        if request.method == 'GET':
            handle = request.args.get('handle')
            if not handle:
                return send_cors_json(EMPTY_DATA)
            doc = col.find_one({'_id': handle})
            if doc:
                doc.pop('_id', None)
                return send_cors_json(doc)
            return send_cors_json(EMPTY_DATA)

        if request.method == 'POST':
            payload = request.get_json(silent=True)
            if not payload or not isinstance(payload, dict):
                return send_cors_json({'error': 'Invalid JSON'}, 400)

            handle = payload.get('handle')
            if not handle or not isinstance(handle, str):
                return send_cors_json({'error': 'handle is required'}, 400)

            action = payload.get('action')
            now = datetime.now(timezone.utc).isoformat()

            if action == 'add_todo':
                problem = payload.get('problem')
                if not problem:
                    return send_cors_json({'error': 'problem is required'}, 400)
                col.update_one(
                    {'_id': handle},
                    {'$push': {'todo': problem}, '$set': {'savedAt': now}},
                    upsert=True
                )
                return send_cors_json({'ok': True})

            if action == 'mark_solved':
                contest_id = payload.get('contestId')
                index = payload.get('index')
                if not contest_id or not index:
                    return send_cors_json({'error': 'contestId and index required'}, 400)
                doc = col.find_one({'_id': handle})
                if not doc:
                    return send_cors_json({'error': 'No upsolve data found'}, 404)
                todo_item = next(
                    (p for p in doc.get('todo', [])
                     if p.get('contestId') == contest_id and p.get('index') == index),
                    None
                )
                if not todo_item:
                    return send_cors_json({'error': 'Problem not found in todo'}, 404)
                solved_item = {**todo_item, 'solvedAt': now}
                col.update_one(
                    {'_id': handle},
                    {
                        '$pull': {'todo': {'contestId': contest_id, 'index': index}},
                        '$push': {'solved': solved_item},
                        '$set': {'savedAt': now}
                    }
                )
                return send_cors_json({'ok': True})

            if action == 'remove_todo':
                contest_id = payload.get('contestId')
                index = payload.get('index')
                if not contest_id or not index:
                    return send_cors_json({'error': 'contestId and index required'}, 400)
                col.update_one(
                    {'_id': handle},
                    {'$pull': {'todo': {'contestId': contest_id, 'index': index}}, '$set': {'savedAt': now}}
                )
                return send_cors_json({'ok': True})

            return send_cors_json({'error': 'Unknown action. Use add_todo, mark_solved, or remove_todo'}, 400)

    except Exception as e:
        print(f'upsolve-data error: {e}')
        return send_cors_json({'error': 'Database error'}, 500)

    return send_cors_json({'error': 'Method Not Allowed'}, 405)
